includes/admin/dashboard-widgets.php

```diff
@@ -152,42 +152,54 @@ function edd_load_dashboard_sales_widget( ) {
 		$payments = edd_get_payments( array( 'number' => 5, 'status' => 'complete' ) );
 
 		if ( $payments ) { ?>
-		<div class="table recent_purchases">
-			<table>
-				<thead>
-					<tr>
-						<td colspan="2">
-							<?php _e( 'Recent Purchases', 'easy-digital-downloads' ); ?>
-							<a href="<?php echo admin_url( 'edit.php?post_type=download&page=edd-payment-history' ); ?>">&nbsp;&ndash;&nbsp;<?php _e( 'View All', 'easy-digital-downloads' ); ?></a>
-						</td>
-					</tr>
-				</thead>
-				<tbody>
-					<?php
-					foreach ( $payments as $payment ) { ?>
-						<tr>
-							<td class="edd_order_label">
-								<a href="<?php echo add_query_arg( 'id', $payment->ID, admin_url( 'edit.php?post_type=download&page=edd-payment-history&view=view-order-details' ) ); ?>">
-									<?php echo get_the_title( $payment->ID ) ?>
-									&mdash; <?php echo $payment->email ?>
-								</a>
-								<?php if ( ! empty( $payment->user_id ) && ( $payment->user_id > 0 ) ) {
-									$user = get_user_by( 'id', $payment->user_id );
-									if ( $user ) {
-										echo "(" . $user->data->user_login . ")";
-									}
-								} ?>
-							</td>
-							<td class="edd_order_price">
-								<a href="<?php echo add_query_arg( 'id', $payment->ID, admin_url( 'edit.php?post_type=download&page=edd-payment-history&view=view-order-details' ) ); ?>">
-									<span class="edd_price_label"><?php echo edd_currency_filter( edd_format_amount( $payment->total ), edd_get_payment_currency_code( $payment->ID ) ); ?></span>
-								</a>
-							</td>
-						</tr>
+		<div class="table recent_orders">
+			<h3><?php esc_html_e( 'Recent Orders', 'easy-digital-downloads' ); ?></h3>
+			<ul>
+			<?php
+			foreach ( $payments as $payment ) {
+				$link = edd_get_admin_url(
+					array(
+						'page' => 'edd-payment-history',
+						'view' => 'view-order-details',
+						'id'   => $payment->ID,
```

Comment:
```suggestion
						'id'   => urlencode( $payment->ID ),
```

Reply:
Thanks, @ashleyfae! I have two questions about your suggestion, please:
- why `urlencode` instead of something like `esc_attr` (or `rawurlencode` because PHPCS says that's preferable)?
- the entire URL is escaped on output with `esc_url` as well, so are both needed?

Reply:
@robincornett `edd_get_admin_url()` uses `add_query_arg()`. You can see the docs on that here: https://developer.wordpress.org/reference/functions/add_query_arg/ Specifically:

> Values are expected to be encoded appropriately with urlencode() or rawurlencode().

So when we are putting together arguments to add to a URL, we should use `urlencode()` or `rawurlencode()` on anything that isn't a hard-coded value ( `'view' => 'view-order-details'` is a hard-coded value so we don't really need it there).

In this particular example, `$payment->ID` is a pretty safe value (it's just an integer), but it's generally good practise to treat everything like it may not be safe/URL-friendly. That's why we would encode it just in case.

As for why not `esc_attr()`, because this particular item at this point of time isn't being inserted into an HTML attribute. It's just not the right place to use it. If this were alt text or something, then we'd use `esc_attr()` around the entire alt text value.

As for why not `rawurlencode()`, we certainly could use that instead, but in this particular case we're working with an integer, so there shouldn't be a practical difference here. The main differences between the two functions are how spaces get encoded and we don't expect this value to contain spaces. There are also some differences with other characters like `~`, but again, we're not expecting any of those characters here. Either function is fine to use because we're dealing with an integer. If it were a string, may contain spaces, etc. then the difference would matter more.

And yes, we should use `urlencode()` in any values we insert into the URL while we're building it, then use `esc_url()` on the entire formed URL right as we display it.